lib/src/data/communication/api/api_exception_mapper.dart

```diff
@@ -0,0 +1,39 @@
+import 'dart:io';
+
+import 'package:allthenews/src/domain/communication/api_exception.dart';
+import 'package:allthenews/src/domain/communication/exception_mapper.dart';
+import 'package:dio/dio.dart';
+
+class ApiExceptionMapper extends ExceptionMapper {
+  @override
+  ApiException toExceptionType(Object error) {
+    if (error is DioError) {
+      switch (error.type) {
+        case DioErrorType.DEFAULT:
+          if (error.error is SocketException) {
+            return ConnectionException();
+          }
+          return UnknownException();
+        case DioErrorType.RESPONSE:
+          return _fromStatusCode(error.response.statusCode);
+        default:
+          return UnknownException();
+      }
+    } else {
+      return UnknownException();
+    }
+  }
+
+  ApiException _fromStatusCode(int code) {
+    switch (code) {
+      case 401:
+        return UnauthorizedException();
+      case 404:
+        return ServerErrorException();
+      case 500:
```

Comment:
i ogólnie to wszystkie statusy błędów zaczynające się od 5 to błędy po stronie serwera więc, nie tylko dla 500, ale dla całego zakresu warto rzucać ServerErrorException :)

Reply:
można, tylko wtedy if else, bo dartowy switch nie obsługuje zakresów

Reply:
można też dodać pozostałe statusy błędów, ale na nasze potrzeby nawet ta 500 jest bardziej jako zapychacz

Reply:
if else zawsze spoko, jak już jest ServerErrorException to myślę, że wydzielenie tych 5* nie boli, a przynajmniej komunikat powie prawdę, że to coś na serwerze, a nie po stronie klienta jak w status codach 4* ;)